torch/nn/utils/attention.py

```diff
@@ -0,0 +1,328 @@
+    """
+
+    UPPER_LEFT = auto()
+    LOWER_RIGHT = auto()
+
+
+class CausalBias(torch.Tensor):
+    """
+    A bias representing causal attention patterns
+
+    This class is used for defining causal (triangular) attention biases.
+
+    .. warning:: This class is a prototype and subject to change.
+    """
+
+    def __init__(self, variant: CausalVariant, seq_len_q: int, seq_len_kv: int):
+        """
+        Initializes the CausalBias instance with a specified variant and sequence lengths.
+
+        Args:
+            variant (CausalVariant): The type of causal bias to use (either UPPER_LEFT or LOWER_RIGHT).
+            seq_len_q (int): The sequence length of the query tensor.
+            seq_len_kv (int): The sequence length of the key/value tensor.
+
+        Raises a warning if the LOWER_RIGHT variant is used with seq_len_q > seq_len_kv, as it may produce NaNs.
+        """
+        assert isinstance(variant, CausalVariant)
+        self.variant = variant
+        self.seq_len_q = seq_len_q
+        self.seq_len_kv = seq_len_kv
+        if seq_len_q > seq_len_kv and variant == CausalVariant.LOWER_RIGHT:
+            warn(
+                "Lower right causal bias will produce NaNs in the output when seq_len_q > seq_len_kv!"
+            )
+
+    @staticmethod
+    def upper_left(seq_len_q: int, seqlen_kv: int) -> "CausalBias":
+        """
+        Creates an upper-left triangular causal bias.
+
+        Args:
+            seq_len_q (int): The sequence length of the query tensor.
+            seq_len_kv (int): The sequence length of the key/value tensor.
+
+        Returns:
+            CausalBias: The upper-left triangular causal bias.
+        """
+        return CausalBias(CausalVariant.UPPER_LEFT, seq_len_q, seqlen_kv)
+
+    @staticmethod
+    def lower_right(seq_len_q: int, seq_len_kv: int) -> "CausalBias":
+        """
+        Creates a lower-right triangular causal bias.
+
+        Args:
+            seq_len_q (int): The sequence length of the query tensor.
+            seq_len_kv (int): The sequence length of the key/value tensor.
+
+        Returns:
+            CausalBias: The lower-right triangular causal bias.
+        """
+        return CausalBias(CausalVariant.LOWER_RIGHT, seq_len_q, seq_len_kv)
+
+    def _upper_left(self, device: torch.device) -> torch.Tensor:
+        """Upper left causal bias"""
+        return torch.tril(
+            torch.ones(self.seq_len_q, self.seq_len_kv, device=device, dtype=torch.bool)
+        )
+
+    def _lower_right(self, device: torch.device) -> torch.Tensor:
+        """Lower right causal bias"""
+        diagonal_offset = self.seq_len_kv - self.seq_len_q
+        return torch.tril(
+            torch.ones(
+                self.seq_len_q, self.seq_len_kv, device=device, dtype=torch.bool
+            ),
+            diagonal=diagonal_offset,
+        )
+
+    def materialize(self, device: Optional[torch.device] = None) -> torch.Tensor:
```

Comment:
Why not to_dense?

Reply:
what if we punt on this for now and make this private as well?